File: examples/grep/grep.c
/*
 * C Grep benchmark - counterpart to grep.mettle
 *
 * Counts lines containing "ERROR" in a 1 MiB log buffer, 200 passes.
 * Uses the same uint64 load + 5-byte mask compare as grep.mettle (memcmp tail only).
 *
 * Build: build.bat (or: gcc -O2 -o grep_c.exe grep.c -lkernel32)
 * Run: grep_c.exe
 */

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <inttypes.h>

#include "../bench_time.h"

#define PATTERN "ERROR"
#define PATTERN_LEN 5
#define PATTERN_U64 0x524F525245ULL
#define PATTERN_MASK 0xFFFFFFFFFFULL
#define GREP_BUF_SIZE 1048576
#define GREP_PASSES 200

static volatile int64_t g_bench_sink;

static int pattern_matches(const unsigned char *buf, int64_t i, int64_t len) {
    if (i + PATTERN_LEN > len) {
        return 0;
    }
    if (i + 8 <= len) {
        uint64_t val;
        memcpy(&val, buf + i, sizeof(val));
        return (int)(((val & PATTERN_MASK) == PATTERN_U64));
    }
    return memcmp(buf + i, PATTERN, PATTERN_LEN) == 0;
}

__attribute__((noinline)) static int64_t grep_count(const unsigned char *buf, int64_t len) {
    int64_t count = 0;
    int found = 0;
    for (int64_t i = 0; i < len; i++) {
        int c = buf[i];
        if (c == '\n') {
            if (found) {
                count++;
            }
            found = 0;
        } else {
            if (!found && c == 'E' && pattern_matches(buf, i, len)) {
                found = 1;
            }
        }
    }
    if (found) {
        count++;
    }
    return count;
}

int main(void) {
    const int64_t buf_size = GREP_BUF_SIZE;
    unsigned char *buf = (unsigned char *)malloc((size_t)buf_size);
    if (!buf) {
        fprintf(stderr, "malloc failed\n");
        return 1;
    }

    const char *block = "INFO 12345\nERROR 12345\nINFO 12345\nINFO 12345\nINFO 1234\n";
    const size_t block_len = 55;
    for (int64_t pos = 0; pos < buf_size; pos += (int64_t)block_len) {
        size_t chunk = (size_t)(buf_size - pos);
        if (chunk > block_len) {
            chunk = block_len;
        }
        memcpy(buf + pos, block, chunk);
    }

    printf("Grep: count lines containing ERROR (1 MiB log buffer)\n");

    int64_t matches = grep_count(buf, buf_size);
    printf("Matches = %" PRId64 "\n", matches);

    const int passes = GREP_PASSES;
    printf("Benchmark: %d passes (grep_count)\n", passes);

    uint64_t t0 = bench_time_us();
    int64_t total = 0;
    for (int p = 0; p < passes; p++) {
        int64_t off = (int64_t)(p & 127);
        total += grep_count(buf + off, buf_size - off);
    }
    uint64_t elapsed_us = bench_time_us() - t0;

    g_bench_sink ^= total;

    printf("Total matches = %" PRId64 "\n", total);
    printf("Time: %" PRIu64 " us\n", elapsed_us);

    uint64_t per_pass_us = elapsed_us / (uint64_t)passes;
    printf("Per pass: ~%" PRIu64 " us\n", per_pass_us);

    free(buf);
    return 0;
}

File: examples/grep/grep.rs
// Rust grep benchmark - same algorithm as grep.c.

#[path = "../bench_time.rs"]
mod bench_time;

use bench_time::bench_time_us;
use std::hint::black_box;

const PATTERN: &[u8] = b"ERROR";
const PATTERN_LEN: i64 = 5;
const PATTERN_U64: u64 = 0x524F_5252_45;
const PATTERN_MASK: u64 = 0xFFFF_FFFF_FF;
const GREP_BUF_SIZE: i64 = 1_048_576;
const GREP_PASSES: i32 = 200;

fn pattern_matches(buf: &[u8], i: i64, len: i64) -> bool {
    if i + PATTERN_LEN > len {
        return false;
    }
    let i = i as usize;
    let len = len as usize;
    if i + 8 <= len {
        let mut val = [0u8; 8];
        val.copy_from_slice(&buf[i..i + 8]);
        let loaded = u64::from_le_bytes(val);
        return (loaded & PATTERN_MASK) == PATTERN_U64;
    }
    &buf[i..i + PATTERN.len()] == PATTERN
}

#[inline(never)]
fn grep_count(buf: &[u8]) -> i64 {
    let len = buf.len() as i64;
    let mut count: i64 = 0;
    let mut found = false;
    for i in 0..len {
        let c = buf[i as usize];
        if c == b'\n' {
            if found {
                count += 1;
            }
            found = false;
        } else if !found && c == b'E' && pattern_matches(buf, i, len) {
            found = true;
        }
    }
    if found {
        count += 1;
    }
    count
}

fn main() {
    let block = b"INFO 12345\nERROR 12345\nINFO 12345\nINFO 12345\nINFO 1234\n";
    let mut buf = vec![0u8; GREP_BUF_SIZE as usize];
    let mut pos: i64 = 0;
    while pos < GREP_BUF_SIZE {
        let chunk = (GREP_BUF_SIZE - pos).min(block.len() as i64) as usize;
        buf[pos as usize..pos as usize + chunk].copy_from_slice(&block[..chunk]);
        pos += chunk as i64;
    }

    println!("Grep: count lines containing ERROR (1 MiB log buffer)");

    let matches = grep_count(&buf);
    println!("Matches = {matches}");
    println!("Benchmark: {GREP_PASSES} passes (grep_count)");

    let t0 = bench_time_us();
    let mut total: i64 = 0;
    for p in 0..GREP_PASSES {
        let off = (p & 127) as usize;
        total += grep_count(&buf[off..]);
        black_box(p);
    }
    let elapsed_us = bench_time_us() - t0;

    println!("Total matches = {total}");
    println!("Time: {elapsed_us} us");
    println!("Per pass: ~{} us", elapsed_us / GREP_PASSES as u64);

    black_box(total);
}
